softhddevice.cpp

```diff
@@ -844,6 +905,9 @@ int cSoftHdDevice::PlayAudio(const uchar *data, int size, uchar id)
 {
 //	LOGDEBUG("device: %s: %p %p %d %d", __FUNCTION__, this, data, size, id);
 
+	if (m_skipstream)
```

Comment:
You could check the state instead. This can be done without synchronization, because the state is only changed by the VDR main thread, and Play*() are only called by it.
```suggestion
	if (m_state == DETACHED)
```

Reply:
I had that one already, but it didn't work, because then the state gets "ATTACHED" before the things in OnEnteringState(ATTACHED). So VDR already already sends data without having cSoftHdAudio ... that ended in a segfault.
Maybe we should change SetSate to
```
if (m_state != newState) {
		LOGDEBUG("device: Preparing to leave state %s", StateToString(m_state));
		OnLeavingState(m_state);
		LOGDEBUG("device: Changing state %s -> %s", StateToString(m_state), StateToString(newState));
		OnEnteringState(newState);
		m_state = newState;
		LOGDEBUG("device: State changed to %s", StateToString(m_state));
	}
```
?

Reply:
Right of the bat, I think that makes more sense generally.

Reply:
Actually the order shouldn't matter if all state changes and Play*() are processed in the VDR main thread. I think my assumption is wrong that Play*() are called from the main thread. In this case there needs to be a mutex around the state and everywhere where the state is accessed.

Reply:
I guess we can assume that Play* is called from the VDR main thread or player thread. I don't think we need to lock anything.
I'll try with the above code. It follows the logic of the state change as it sets m_state not until we really reached it.

Reply:
Changing that will follow the logic but we open the next problem:
When entering DETACHED, the first we want it to skip data in Play* - But DETACHED is set not until OnEnteringState has finished then. So, though it works for entering ATTACHED it doesn't work for entering DETACHED anymore. I'd suggest to use m_skipStream for now.

Reply:
So i have to correct myself. We need to lock the state machine, because when we use svdrp (which is a separate thread) we run into a problem if we are in between a state change and want to set a new state....

Reply:
https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/pull/113/commits/3230c8c1da537b1c4c07ffe112792955d81403c8 should fix that

Reply:
Since there are three threads, now, changing the state (main, player, svdrp), we need the mutex here, too. Otherwise, the decoder could be closed, while we are in Play*().

Reply:
You mean locking the whole PlayVideo and PlayAudio to protect the m_skipStream?
I guess, we can use the same mutex for both, because PlayAudio and PlayVideo are called from the same VDR thread?

We have to look into PlayVideoPkts/PlayAudioPkts afterwards, though.

Reply:
Not only to protect m_skipStream, but also to protect the logic inside the decoders. Otherwise a thread can change/close the decoder while another thread sends data to it via Play*(). I'd say we *must* use the same mutex for both play methods (as for the state mutex).

Reply:
Let me push sth...